lib/omniauth/strategies/apple.rb

```diff
@@ -1,55 +1,87 @@
+# frozen_string_literal: true
+
 require 'omniauth-oauth2'
 
 module OmniAuth
   module Strategies
     class Apple < OmniAuth::Strategies::OAuth2
-
-      attr_reader :id_token
-      args %i[client_id team_id key_id pem]
-
       option :name, 'apple'
-      option :client_options, {
-        site: 'https://appleid.apple.com',
-        authorize_url: '/auth/authorize',
-        token_url: '/auth/token',
-      }
 
-      uid { id_token['sub'] }
+      option :client_options,
+             site: 'https://appleid.apple.com',
+             authorize_url: '/auth/authorize',
+             token_url: '/auth/token'
+
+      uid { id_info['sub'] }
 
       info do
-        { sub: id_token['sub'] }
+        {
+          sub: id_info['sub'],
+          email: user_info.dig('email'),
+          first_name: user_info.dig('name', 'firstName'),
+          last_name: user_info.dig('name', 'lastName'),
+          extra: {
+            raw_info: id_info.merge(user_info)
+          }
+        }
       end
 
       def client
         ::OAuth2::Client.new(options.client_id, client_secret, deep_symbolize(options.client_options))
       end
 
       def callback_url
-        full_host + script_name + callback_path
+        options[:redirect_uri] || (full_host + script_name + callback_path)
       end
 
-      def build_access_token
-        _access_token = super
-        @id_token = ::JSON::JWT.decode(_access_token.params['id_token'], :skip_verification)
```

Comment:
@nhosoya What was the reason for intercepting the `build_access_token` method here instead of reading the `access_token.params` later?